# // RANDOM FOREST REGRESSOR //

#Importing necessary libraries for data preprocessing

In [1]:
import pandas as pd #For reading the dataset and to perform framing functions if needed
import numpy as np  #linear algebra function
import matplotlib.pyplot as plt #for visualizing the data
import seaborn as sns #for data distribution
import warnings
warnings.filterwarnings("ignore")

Matplotlib is building the font cache; this may take a moment.


#Importing the dataset
I have used a simple 'weather status dataset' to perform a decision tree classifier

In [2]:
#---> Replace the path name for your path name
tips_data=pd.read_csv('C:/Users/ADMIN/Documents/Data science algorithms/tips.csv')

#Analysing and preprocessing the dataset

In [3]:
tips_data.head() #---> First 5 rows

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [4]:
tips_data.nunique()

total_bill    229
tip           123
sex             2
smoker          2
day             4
time            2
size            6
dtype: int64

In [5]:
tips_data.shape

(244, 7)

In [6]:
summary={'Columns':tips_data.columns,
        'Unique Values':[tips_data[col].nunique() for col in tips_data.columns],
        'Details':[tips_data[col].unique() for col in tips_data.columns]}

In [7]:
details=pd.DataFrame(summary)
details

Columns  Unique Values  \
0  total_bill            229   
1         tip            123   
2         sex              2   
3      smoker              2   
4         day              4   
5        time              2   
6        size              6   

                                             Details  
0  [16.99, 10.34, 21.01, 23.68, 24.59, 25.29, 8.7...  
1  [1.01, 1.66, 3.5, 3.31, 3.61, 4.71, 2.0, 3.12,...  
2                                     [Female, Male]  
3                                          [No, Yes]  
4                              [Sun, Sat, Thur, Fri]  
5                                    [Dinner, Lunch]  
6                                 [2, 3, 4, 1, 6, 5]

In [8]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
label_encode=LabelEncoder()
tips_data['sex_encode']=label_encode.fit_transform(tips_data['sex'])
tips_data['smoker_encode']=label_encode.fit_transform(tips_data['smoker'])
tips_data['day_encode']=label_encode.fit_transform(tips_data['day'])
tips_data['time_encode']=label_encode.fit_transform(tips_data['time'])

In [9]:
x=tips_data[['total_bill','sex_encode','smoker_encode','day_encode','time_encode','size']]
x.head()

total_bill  sex_encode  smoker_encode  day_encode  time_encode  size
0       16.99           0              0           2            0     2
1       10.34           1              0           2            0     3
2       21.01           1              0           2            0     3
3       23.68           1              0           2            0     2
4       24.59           0              0           2            0     4

In [10]:
y=tips_data.tip
y.head()

0    1.01
1    1.66
2    3.50
3    3.31
4    3.61
Name: tip, dtype: float64

#Splitting the dataset into training and testing dataset

In [11]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

#Training the model

In [12]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV
regressor=RandomForestRegressor()

In [13]:
paramaters={'criterion':['squared_error','absolute_error','friedman_mse','poisson']}

In [14]:
forest_regressor=GridSearchCV(regressor,paramaters,scoring='accuracy',cv=5)
forest_regressor.fit(x_train,y_train)

GridSearchCV(cv=5, estimator=RandomForestRegressor(),
             param_grid={'criterion': ['squared_error', 'absolute_error',
                                       'friedman_mse', 'poisson']},
             scoring='accuracy')

In [15]:
forest_regressor.best_params_

{'criterion': 'squared_error'}

In [16]:
predictions=forest_regressor.predict(x_train)
predictions.flatten()

array([2.5006, 2.4763, 3.4572, 4.8275, 2.0338, 3.1827, 2.0085, 3.818 ,
       2.2922, 2.833 , 2.8239, 4.8241, 3.1444, 5.0393, 2.7811, 2.4119,
       3.7278, 2.2057, 1.5759, 3.3408, 3.6237, 2.5487, 3.2044, 2.8755,
       2.9803, 2.7003, 3.1705, 3.821 , 4.4628, 3.2539, 4.8364, 1.6904,
       3.9492, 2.9453, 1.7118, 3.4155, 2.1941, 2.866 , 1.8084, 3.341 ,
       4.109 , 2.8126, 4.8118, 1.5351, 3.5458, 3.147 , 2.3895, 4.5146,
       3.2663, 3.8071, 3.2199, 2.7432, 3.5894, 2.8549, 3.318 , 2.469 ,
       2.166 , 1.567 , 2.3268, 5.5506, 2.0507, 1.7825, 3.4229, 4.0884,
       4.4947, 2.2403, 2.1545, 1.7781, 3.064 , 3.3823, 5.2139, 8.3163,
       2.3004, 3.3286, 2.426 , 4.7484, 2.0966, 2.5013, 1.8421, 2.4675,
       1.927 , 5.034 , 4.5576, 2.072 , 2.5764, 3.6601, 2.3839, 2.7474,
       6.2203, 3.0895, 2.1186, 1.8836, 1.3638, 1.9561, 1.9879, 3.3693,
       2.075 , 3.0302, 2.7286, 5.265 , 3.2984, 2.3866, 2.0335, 6.2242,
       3.8102, 2.9383, 3.2004, 4.588 , 4.0662, 5.5257, 3.4029, 4.3348,
      

In [17]:
from sklearn.metrics import r2_score
score=r2_score(y_train,predictions)
print(score*100)

90.78747613327454


#This regressor model has produced a accuracy score of 91%.